File: regions/system/library/rust/migration/src/analyzer.rs
use sqlparser::ast::{AlterTableOperation, Statement};
use sqlparser::dialect::GenericDialect;
use sqlparser::parser::Parser;

#[derive(Debug, Clone, PartialEq)]
pub enum BreakingChange {
    ColumnDropped {
        table: String,
        column: String,
    },
    ColumnTypeChanged {
        table: String,
        column: String,
        from: String,
        to: String,
    },
    TableDropped {
        table: String,
    },
    NotNullAdded {
        table: String,
        column: String,
    },
    ColumnRenamed {
        table: String,
        from: String,
        to: String,
    },
}

impl std::fmt::Display for BreakingChange {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ColumnDropped { table, column } => {
                write!(f, "Column {table}.{column} dropped")
            }
            Self::ColumnTypeChanged {
                table,
                column,
                from,
                to,
            } => write!(
                f,
                "Column {table}.{column} type changed from {from} to {to}"
            ),
            Self::TableDropped { table } => write!(f, "Table {table} dropped"),
            Self::NotNullAdded { table, column } => {
                write!(f, "NOT NULL added to {table}.{column}")
            }
            Self::ColumnRenamed { table, from, to } => {
                write!(f, "Column {table}.{from} renamed to {to}")
            }
        }
    }
}

/// SQLから破壊的変更を検出する
pub fn detect_breaking_changes(sql: &str) -> Vec<BreakingChange> {
    let dialect = GenericDialect {};
    let statements = match Parser::parse_sql(&dialect, sql) {
        Ok(stmts) => stmts,
        Err(_) => return Vec::new(),
    };

    let mut changes = Vec::new();

    for stmt in &statements {
        match stmt {
            Statement::Drop {
                object_type, names, ..
            } => {
                use sqlparser::ast::ObjectType;
                if *object_type == ObjectType::Table {
                    for name in names {
                        changes.push(BreakingChange::TableDropped {
                            table: name.to_string(),
                        });
                    }
                }
            }
            Statement::AlterTable {
                name, operations, ..
            } => {
                let table = name.to_string();
                for op in operations {
                    match op {
                        AlterTableOperation::DropColumn { column_name, .. } => {
                            changes.push(BreakingChange::ColumnDropped {
                                table: table.clone(),
                                column: column_name.to_string(),
                            });
                        }
                        AlterTableOperation::AlterColumn {
                            column_name, op, ..
                        } => {
                            use sqlparser::ast::AlterColumnOperation;
                            match op {
                                AlterColumnOperation::SetNotNull => {
                                    changes.push(BreakingChange::NotNullAdded {
                                        table: table.clone(),
                                        column: column_name.to_string(),
                                    });
                                }
                                AlterColumnOperation::SetDataType {
                                    data_type, ..
                                } => {
                                    changes.push(BreakingChange::ColumnTypeChanged {
                                        table: table.clone(),
                                        column: column_name.to_string(),
                                        from: "unknown".to_string(),
                                        to: data_type.to_string(),
                                    });
                                }
                                _ => {}
                            }
                        }
                        AlterTableOperation::RenameColumn {
                            old_column_name,
                            new_column_name,
                        } => {
                            changes.push(BreakingChange::ColumnRenamed {
                                table: table.clone(),
                                from: old_column_name.to_string(),
                                to: new_column_name.to_string(),
                            });
                        }
                        _ => {}
                    }
                }
            }
            _ => {}
        }
    }

    changes
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_drop_table() {
        let sql = "DROP TABLE users;";
        let changes = detect_breaking_changes(sql);
        assert_eq!(changes.len(), 1);
        assert_eq!(
            changes[0],
            BreakingChange::TableDropped {
                table: "users".to_string()
            }
        );
    }

    #[test]
    fn test_detect_drop_column() {
        let sql = "ALTER TABLE users DROP COLUMN email;";
        let changes = detect_breaking_changes(sql);
        assert_eq!(changes.len(), 1);
        assert_eq!(
            changes[0],
            BreakingChange::ColumnDropped {
                table: "users".to_string(),
                column: "email".to_string()
            }
        );
    }

    #[test]
    fn test_detect_set_not_null() {
        let sql = "ALTER TABLE users ALTER COLUMN email SET NOT NULL;";
        let changes = detect_breaking_changes(sql);
        assert_eq!(changes.len(), 1);
        assert_eq!(
            changes[0],
            BreakingChange::NotNullAdded {
                table: "users".to_string(),
                column: "email".to_string()
            }
        );
    }

    #[test]
    fn test_detect_type_change() {
        let sql = "ALTER TABLE users ALTER COLUMN age SET DATA TYPE BIGINT;";
        let changes = detect_breaking_changes(sql);
        assert_eq!(changes.len(), 1);
        match &changes[0] {
            BreakingChange::ColumnTypeChanged { table, column, to, .. } => {
                assert_eq!(table, "users");
                assert_eq!(column, "age");
                assert_eq!(to, "BIGINT");
            }
            other => panic!("Expected ColumnTypeChanged, got {:?}", other),
        }
    }

    #[test]
    fn test_detect_rename_column() {
        let sql = "ALTER TABLE users RENAME COLUMN old_name TO new_name;";
        let changes = detect_breaking_changes(sql);
        assert_eq!(changes.len(), 1);
        assert_eq!(
            changes[0],
            BreakingChange::ColumnRenamed {
                table: "users".to_string(),
                from: "old_name".to_string(),
                to: "new_name".to_string()
            }
        );
    }

    #[test]
    fn test_no_breaking_changes() {
        let sql = "ALTER TABLE users ADD COLUMN email TEXT;";
        let changes = detect_breaking_changes(sql);
        assert!(changes.is_empty());
    }

    #[test]
    fn test_display_formatting() {
        let change = BreakingChange::ColumnDropped {
            table: "users".to_string(),
            column: "email".to_string(),
        };
        assert_eq!(change.to_string(), "Column users.email dropped");
    }

    #[test]
    fn test_invalid_sql_returns_empty() {
        let changes = detect_breaking_changes("NOT VALID SQL AT ALL !!!");
        assert!(changes.is_empty());
    }
}

File: regions/system/library/typescript/migration/src/analyzer.ts
export type BreakingChange =
  | { type: 'columnDropped'; table: string; column: string }
  | { type: 'columnTypeChanged'; table: string; column: string; from: string; to: string }
  | { type: 'tableDropped'; table: string }
  | { type: 'notNullAdded'; table: string; column: string }
  | { type: 'columnRenamed'; table: string; from: string; to: string };

export function formatBreakingChange(change: BreakingChange): string {
  switch (change.type) {
    case 'columnDropped':
      return `Column ${change.table}.${change.column} dropped`;
    case 'columnTypeChanged':
      return `Column ${change.table}.${change.column} type changed from ${change.from} to ${change.to}`;
    case 'tableDropped':
      return `Table ${change.table} dropped`;
    case 'notNullAdded':
      return `NOT NULL added to ${change.table}.${change.column}`;
    case 'columnRenamed':
      return `Column ${change.table}.${change.from} renamed to ${change.to}`;
  }
}

/**
 * Detect breaking changes in SQL using regex-based parsing.
 * Returns an empty array for invalid/unrecognized SQL.
 */
export function detectBreakingChanges(sql: string): BreakingChange[] {
  const statements = sql
    .split(';')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);

  const changes: BreakingChange[] = [];

  for (const stmt of statements) {
    const dropTable = stmt.match(/^DROP\s+TABLE\s+(?:IF\s+EXISTS\s+)?(\S+)/i);
    if (dropTable) {
      changes.push({ type: 'tableDropped', table: dropTable[1] });
      continue;
    }

    const dropColumn = stmt.match(/^ALTER\s+TABLE\s+(\S+)\s+DROP\s+COLUMN\s+(?:IF\s+EXISTS\s+)?(\S+)/i);
    if (dropColumn) {
      changes.push({ type: 'columnDropped', table: dropColumn[1], column: dropColumn[2] });
      continue;
    }

    const setNotNull = stmt.match(/^ALTER\s+TABLE\s+(\S+)\s+ALTER\s+COLUMN\s+(\S+)\s+SET\s+NOT\s+NULL/i);
    if (setNotNull) {
      changes.push({ type: 'notNullAdded', table: setNotNull[1], column: setNotNull[2] });
      continue;
    }

    const typeChange = stmt.match(/^ALTER\s+TABLE\s+(\S+)\s+ALTER\s+COLUMN\s+(\S+)\s+(?:SET\s+DATA\s+)?TYPE\s+(\S+)/i);
    if (typeChange) {
      changes.push({
        type: 'columnTypeChanged',
        table: typeChange[1],
        column: typeChange[2],
        from: 'unknown',
        to: typeChange[3],
      });
      continue;
    }

    const renameColumn = stmt.match(/^ALTER\s+TABLE\s+(\S+)\s+RENAME\s+COLUMN\s+(\S+)\s+TO\s+(\S+)/i);
    if (renameColumn) {
      changes.push({
        type: 'columnRenamed',
        table: renameColumn[1],
        from: renameColumn[2],
        to: renameColumn[3],
      });
      continue;
    }
  }

  return changes;
}
